### DISCUSSION TITLE
Découpage de code : mécanismes

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#184] [Ideas]
Actuellement un utilisateur de la librairie `publicodes` importe l'ensemble des mécanismes dans son application Javascript.

Dans la lignée de https://github.com/betagouv/publicodes/discussions/183, on pourra ajouter un import ES module pour chaque mécanisme utilisé dans un fichier de règles publicodes :

```js
// impot.publicodes.yaml
import 'publicodes/mechanisms/somme'
import 'publicodes/mechanisms/barème'
import 'publicodes/mechanisms/abattement'
```

Cela permettra d'ajouter autant de mécanismes voulu dans la librairie standard sans alourdir le poids du javascript envoyé aux utilisateurs finaux si l'application n'utilise pas certains mécanismes fournis.

Côté organisation du code cela veut aussi dire qu'il faut que les mécanismes puisse “s'enregistrer” comme des “plugins” et non plus être importés par le point d'entrée du moteur. Cela devrait permettre à des utilisateurs tiers de développer leurs propres mécanismes.

**Comment by johangirod** [comment]
Dans la lignée de #23, on peut conclure qu'il y a trois types de mécanismes : 
- Ceux de base, qui sont les briques de base du langage 
- Ceux qui peuvent être définis avec d'autres mécanismes, et qui n'ont pas besoin d'être importés : ce sont des fonctions supplémentaires `core`
- Ceux qui sont définis dans des paquets séparés, soit "officiels", soit définis par l'utilisateur

> Cela devrait permettre à des utilisateurs tiers de développer leurs propres mécanismes

Je pense que cela devrait être vu comme la définition d'une fonction, et inclus dans le langage publicodes. Il faut que "mécanisme custom" = "définition de fonction". D'où mon envie de renommer "mécanisme" en "fonction" #57

**Comment by mquandalle** [comment]
Une idée qui me trotte dans la tête depuis déjà longtemps, utiliser les mécanismes directement comme des fonctions : 


```js
// au lieu de
engine.evaluate({'est applicable': rule})

// importer le mecanisme sous forme de fonction
import { est_applicable } from "publicodes/mechanisms"
engine.evaluate(est_applicable(rule));
```

C'est très proche et je pense que toutes les fonctionnalités que l'on peut imaginer dans un cas peuvent aussi être implémenté dans l'autre, mais l'approche « mécanisme utilisé sous forme de fonction » rend plus “native” l'intégration avec l'écosystème JS : 
- typage TS de l'évaluation (ici booléen)
- tree-shaking des mécanismes non utilisés
- code plus facilement minifiable
- création de mécanisme par les utilisateurs qui appellent d'autres mécanismes

Comme JSX ou d'autres librairies qui fonctionnent avec du calcul symbolique, `est_applicable` n'a pas pour but de déterminer l'applicabilité, mais seulement de décomposer cette opération en briques de base. C'est ensuite `engine.evaluate` qui fournit l'implémentation concrète de ces briques de base.

C'est proche de ce qu'on fait déjà, il s'agit essentiellement d'écrire (ou de compiler le yaml en) `somme(ref('a'), valeur(4))` au lieu de `{ somme: ['a', 4] }`.
